src/main/java/opendcs/dao/DaoBase.java

```diff
@@ -56,578 +56,671 @@
+                catch (SQLException ex)
+                {
+                    String msg = String.format(
+                            "Attempting to set parameter of type '%'"
+                        + ". Please open an issue on the project page with this error message.",
+                        param.getClass().getName());
+                    throw new SQLException(msg, ex);
+                }
+            }
+            index++;
+        }
+    }
+
+    /**
+     * Prepare and run a statement with the given parameters.
+     * Each element of the result set is passed to the consumer
+     *
+     * System will make best effort to automatically convert datatypes.
+     * @see DaoBase#withConnection(ConnectionConsumer) for thread safety
+     * @param query SQL Query string with ? for bind variables
+     * @param consumer User provided function to use the result set. @see opendcs.util.functional.ResultSetConsumer
+     * @param parameters Variables for the query
+     * @throws SQLException
+     */
+    public void doQuery(String query, ResultSetConsumer consumer, Object... parameters) throws SQLException
+    {
+        withStatement(query, (stmt) -> {
+            try (ResultSet rs = stmt.executeQuery();)
+            {
+                while(rs.next())
+                {
+                    consumer.accept(rs);
+                }
+            }
+        },parameters);
+    }
+
+    /**
+     * perform an update or insert operations with given parameters.
+     * @param query SQL query passed directly to prepareStatement
+     * @param args variables to bind on the statement in order.
+     * @return number of rows affected
+     * @throws SQLException anything goes wrong talking to the database
+     */
+    public int doModify(String query, Object... args) throws SQLException
+    {
+        int result[] = new int[1];
+        result[0] = 0;
+        withStatement(query, (stmt) -> {
+            result[0] = stmt.executeUpdate();
+        },args);
+        return result[0];
+    }
+
+    /**
+     * Perform a repeated SQL query with the given list of values and a bindingFunction
+     * @param <ValueType> Object we are binding to the query
+     * @param query The query to use repeatedly
+     * @param bindingFunction A function that takes the ValueType and creates a Object[] list in bind variable order
+     * @param values List of values that will be used.
+     * @param batchSize How many elements of the list to execute for each batch.
+     * @throws SQLException
+     */
+    public <ValueType> void doModifyBatch(String query, ThrowingFunction<ValueType, Object[] , SQLException> bindingFunction, List<ValueType> values, int batchSize) throws SQLException
+    {
+        withStatement(query, (stmt) -> {
+            int count = 0;
+            for (ValueType v: values)
+            {
+                bind(stmt, bindingFunction.accept(v));
+                stmt.addBatch();
+                if(count++ % batchSize == 0)
```

Comment:
```suggestion
                if(++count % batchSize == 0)
```
With count++  the batch will run on the first iteration because 0%batchSize ==0